### DISCUSSION TITLE
How should Pascal's forward routine declarations be handled?

### DISCUSSION THREAD

**Discussion by texdraft** [shreevatsa/webWEB#11] [Q&A]
Pascal has a bizarre convention for declaring routines (procedures/functions) before they are defined: Instead of a block after the routine heading, you write the identifier `forward`. For example, section 78 of TeX has

```pascal
procedure normalize_selector; forward;
procedure get_token; forward;
procedure term_input; forward;
procedure show_context; forward;
procedure begin_file_reading; forward;
procedure open_log_file; forward;
procedure close_files_and_terminate; forward;
procedure clear_for_error_prompt; forward ;
procedure give_err_help; forward;
```

And then in section 92 we have the actual definition of `normalize_selector`, in section 365 we have the actual definition of `get_token`, and so on.

If a routine takes arguments, then they must be specified *only* at the point of the `forward` declaration. You can say

```pascal
function I_take_arguments(x: integer); forward;
⋮
function I_take_arguments;
  …;
```

Knuth has some strong words to say about this aspect of the language (section 693 of TeX):

> The inelegant introduction of `show_info` in the code above seems better than the alternative of using
Pascal’s strange `forward` declaration for a procedure with parameters. The Pascal convention about dropping
parameters from a post-`forward` procedure is, frankly, so intolerable to the author of TEX that he would
rather stoop to communication via a global temporary variable. (A similar stoopidity occurred with respect
to `hlist_out` and `vlist_out` above, and it will occur with respect to `mlist_to_hlist` below.)

----

My question is this: How should `forward` declarations be handled in hypertext output? For example, what should happen if you click on a use of `normalize_selector`? Should there be a distinction between clicking on a `forward`-declared routine before it is defined (e.g., `normalize_selector` in section 93) versus clicking on it after it is defined (e.g., `normalize_selector` in section 863)?

In standard C, where forward declarations are more reasonable, parameter information must be repeated both at the point of declaration and at the point of definition. (This isn't true for pre-prototype C, in which `float f();` is a valid declaration for `float f(x,y,z) float x,y,z;`.) Thus you are losing no information by jumping always to the definition; this is what Woboq does. But in Pascal, to link to the definition would mean to miss out on the details of parameters, and to link to the declaration would mean to miss out on the body. (Dropping parameters is a non-issue in Knuth's programs, because he dislikes the convention so strongly. However, I would like to use my program for generating cross-referenced listings of other Pascal programs, which may have `forward`-declared routines with parameters.)

One idea is to have all calls link to the definition, but the routine name in the definition links to the point of `forward` declaration. There should be a way to get from the declaration to the definition&mdash;perhaps by clicking on the word `forward`?

**Comment by shreevatsa** [comment]
Interesting question :-) A couple of thoughts:

1.  The following is what Google's code browser (https://cs.opensource.google/ or https://source.chromium.org/chromium) does (not for Pascal which I don't think it supports, but generally speaking):

    -   When you click on the place where a function (or variable) is used / referenced, it takes you to the place where it is defined (or declared, possibly),
    -   If you are already there, then when you click on it, a pane opens up in the bottom of the window, showing all references etc.

    For example, try clicking on function names starting say [here](https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/renderer/core/paint/svg_root_painter.cc;l=29-34;drc=77a9a09de0688ca449f5333a305ceaf3f36b6daf) (nothing special about this file; it's just a link I had from earlier, from [this bug](https://crbug.com/848748)).

    Doing something like this will also address one minor awkwardness I felt when clicking around in the Lisp source at https://texdraft.github.io/ or the Pascal at https://texdraft.github.io/common-words.html (I haven't tried Woboq much) -- it's a bit surprising to click on something and arrive at the same place, especially if the page jumps around a bit (because the clicked element is scrolled to the top of the page).

2.  As implementing an elaborate pane like this may be too much work, something simple that works for this Pascal forward-reference case is like what you mention in the last line: have the forward declaration and the definition point to each other (so that clicking on one takes you to the other), and point all other uses to either the declaration or the definition (pick one). That way a reader can always get from any usage site to either the declaration or the definition, in at most two clicks.

**Reply by texdraft** [reply]
The motivation behind making identifiers link to themselves in declarations/definitions was to provide an easy way of getting a &ldquo;permalink&rdquo; to the declaration/definition you're reading, if no references to it are nearby. I also thought it is what Woboq does, but I was wrong.

In fact, every code browser I've looked at does the exact same thing when clicking on an identifier at its point of introduction: They take you to a page (or bring up a panel) that shows all the references and so on. (Woboq is a slight exception because you actually can't *click* on the identifier; you have to hover over it.)

I think this is a better move than what I did. First, it will be what users of other tools expect. Second, it is more general, since the page/panel can contain all sorts of information about the identifier. (Although it raises some issues regarding labels.)

As for `forward`, perhaps there could be a little arrow after the `forward` that takes you to the definition. For example, section 78's Pascal part might begin thus (the links are there just to give an idea of what it would look like):

> **procedure** [*normalize_selector*](https://example.com); *forward* [&rarr;](https://example.com);

Clicking on *normalize_selector* takes you to the information for the symbol, and clicking on the arrow takes you to section 92, which has the procedure's definition:

> **procedure** [*normalize_selector*](https://example.com);  
> &nbsp;&nbsp;**begin** &hellip; **end**;

Now we have a new question: What should clicking on *normalize_selector* do here? Taking you to section 78 seems a little inconsistent, but otherwise there would be no easy way to go to the `forward` declaration. Then again, whether it's useful to link to the `forward` declaration is debatable.
